### PR TITLE
[SPARK-28525][DEPLOY] Allow Launcher to be applied Java options

### PR DESCRIPTION
**Pull request by sarutak**
Launcher is implemented as a Java application and sometimes I'd like to apply Java options.
One situation I have met is the time I try to attach debugger to Launcher.

Launcher is launched from bin/spark-class but there is no room to apply Java options.

```
build_command() {
  "$RUNNER" -Xmx128m -cp "$LAUNCH_CLASSPATH" org.apache.spark.launcher.Main "$@"
  printf "%d\0" $?
}
```

Considering that it's not so many times to apply Java options to Launcher,  one compromise would just modify spark-class by user like as follows.

```
build_command() {
  "$RUNNER" -Xmx128m $SPARK_LAUNCHER_OPTS -cp "$LAUNCH_CLASSPATH" org.apache.spark.launcher.Main "$@"
  printf "%d\0" $?
}
```

But it doesn't work when any text related to Java options is output to standard output because whole output is used as command-string for spark-shell and spark-submit in current implementation.

One example is jdwp. When apply agentlib option to use jdwp for debug, we will get output like as follows.

```
Listening for transport dt_socket at address: 9876
```

The output shown above is not a command-string so spark-submit and spark-shell will fail.
To enable Java options for Launcher, we need treat command-string and others.

I changed launcher/Main.java and bin/spark-class to print separator-character and treat it.

## How was this patch tested?

Tested manually using Spark Shell with / without LAUNCHER_JAVA_OPTIONS like as follows.

```
SPARK_LAUNCHER_OPTS="-agentlib:jdwp=transport=dt_socket,suspend=y,address=localhost:9876,server=y" bin/spark-shell
```

### CODE REVIEW

**Comment by sarutak** [comment]
retest this please.

**Comment by xuanyuanking** [comment]
Is the `SPARK_SUBMIT_OPTS` parameter cover your requirement?

**Comment by sarutak** [comment]
@xuanyuanking Launcher and SparkSubmit are run on different JVM process so we can't use `SPARK_SUBMIT_OPTS` for this purpose.

File: bin/spark-class [hunk 1]
```diff
@@ -68,15 +68,26 @@ fi
...
 }
 
 # Turn off posix mode since it does not allow process substitution
 set +o posix
 CMD=()
-while IFS= read -d '' -r ARG; do
-  CMD+=("$ARG")
+DELIM=$'\n'
+while IFS= read -d "$DELIM" -r ARG; do
+  if [ -n "$CMD_START_FLAG" ]; then
```
Comment (vanzin) [inline bin/spark-class]: Better to initialize `CMD_START_FLAG` to an empty string before the loop.

Or you could use `DELIM` for the same effect.

**Comment by kiszk** [comment]
Do we want to update `configuration.md`?

File: bin/spark-class (hunk 1 above)
Comment (sarutak) [reply bin/spark-class]: Thanks. I'll initialize the flag.

**Comment by sarutak** [comment]
@kiszk I think this environment variable is for developers so how about update `spark-env.sh.template` rather?

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
Lot's of flaky tests...

retest this please.

**Comment by sarutak** [comment]
retest this please.

**Review by kiszk — APPROVED** [review]

**Comment by vanzin** [comment]
Merging to master. (Hopefully nobody will ask for this on Windows...)

**Comment by maropu** [comment]
I fetched this commit and my spark-shell seemed to be broken... do I miss something?
```
(py27) maropu@~/Repositories/spark/spark-master: (master $%>)$./bin/spark-shell
NOTE: SPARK_PREPEND_CLASSES is set, placing locally compiled Spark classes ahead of assembly.
/Users/maropu/Repositories/spark/spark-master/bin/spark-class: line 96: CMD: bad array subscript
head: illegal line count -- -1
```

**Comment by kiszk** [comment]
In my environment (Ubuntu 18.04), it works well.

```
$ git log | head -1
commit dba4375359a2dfed1f009edc3b1bcf6b3253fe02
$  ./bin/spark-shell 
19/07/31 13:01:44 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
Spark context Web UI available at http://s7fp1.trl.ibm.com:4040
Spark context available as 'sc' (master = local[*], app id = local-1564545717751).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.0.0-SNAPSHOT
      /_/
         
Using Scala version 2.12.8 (OpenJDK 64-Bit Server VM, Java 1.8.0_212)
Type in expressions to have them evaluated.
Type :help for more information.

scala>
```

**Comment by maropu** [comment]
oh, I see. I'll check again later, thx!

**Comment by maropu** [comment]
It seems my bash is too old;
```
$bash --version
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin16)
Copyright (C) 2007 Free Software Foundation, Inc.
```

**Comment by sarutak** [comment]
It's strange. I use bash of same version for you.

```
scala> :quit
$ bash --version
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin15)
Copyright (C) 2007 Free Software Foundation, Inc.
```

But the version of macOS is different. Is this the cause?

**Comment by vanzin** [comment]
> /Users/maropu/Repositories/spark/spark-master/bin/spark-class: line 96: CMD: bad array subscript

That's also interesting because that line hasn't changed.

**Comment by vanzin** [comment]
Takeshi, I can see that happening if you have the new script with an old build of the launcher jar. Otherwise, the logic seems right (and works for me).

**Comment by maropu** [comment]
That's it! You're right; I rebuilt the package, then the problem has gone. Thx!

**Comment by alfozan** [comment]
I also had the same issue as @maropu 
```
/Users/maropu/Repositories/spark/spark-master/bin/spark-class: line 96: CMD: bad array subscript
```
**Solution: rebuild**
```
 ./build/mvn -Pnetlib-lgpl -Phadoop-2.4 -Phive -Phive-thriftserver -DskipTests clean package
```

**Comment by jahidhasanlinix** [comment]
I have this problem when trying to run spark shell, my java env path seems ok, but its still giving me this error>:
echo $JAVA_HOME
/usr/bin/java:/usr/lib/jvm/java-11-openjdk-amd64

Here's the error while trying to run the shell
#./spark-shell 
/usr/local/spark/spark-3.2.0-bin-hadoop3.2/bin/spark-class: line 71: /usr/bin/java:/usr/lib/jvm/java-11-openjdk-amd64/bin/java: No such file or directory
/usr/local/spark/spark-3.2.0-bin-hadoop3.2/bin/spark-class: line 96: CMD: bad array subscript

**Comment by sarutak** [comment]
@jahidhasanlinix 
Maybe, your `JAVA_HOME` is wrong.
Please set `/usr/lib/jvm/java-11-openjdk-amd64` to `JAVA_HOME`.

**Pull request closed without merging**